load-test/bin/generate-pupil-load-test-data.js

```diff
@@ -1,101 +1,107 @@
-      VALUES (999, @estabCode, @name, @urn, @dfeNumber)`
-      await sqlService.query(sql, params)
+
+    let schools, numSchools, pupilsPerSchool, pupilsRemainder, params
+
+    schools = await sqlService.query(`SELECT id FROM ${sqlService.adminSchema}.[school]`)
+    numSchools = schools.length
+
+    // When generating less pupils than the total number of schools, slice
+    // the schools array to the number of pupils and set 1 pupil per school
+    pupilsPerSchool = Math.floor(numPupils / numSchools)
+    if (pupilsPerSchool < 1) {
+      pupilsPerSchool = 1
+      schools = schools.slice(0, numPupils)
+      numSchools = schools.length
+      pupilsRemainder = 0
+    } else {
+      pupilsRemainder = numPupils - (pupilsPerSchool * numSchools)
     }
-    const testSchools = await schoolDataService.sqlFindByDfeNumbers(dfeNumbers)
-    // Generate additional pupils for these schools and pins
-    winston.info('Generating pupils within schools and pins...')
-    await Promise.all(testSchools.map(async s => {
-      const params = [
+
+    winston.info(`Generating ${numPupils} pupils across ${numSchools} schools`)
+
+    for (let i = 0; i < numSchools; i++) {
+      let school = schools[i]
+      let totalPupils = pupilsPerSchool
+      if (i < pupilsRemainder) {
+        totalPupils += 1
+      }
+      params = [
         {
           name: 'schoolId',
-          value: s.id,
-          type: TYPES.NVarChar
+          value: school.id,
+          type: TYPES.Int
         },
         {
           name: 'dateOfBirth',
           value: randomDob(),
           type: TYPES.DateTimeOffset
+        },
+        {
+          name: 'pinExpiresAt',
+          value: pinExpireyDate,
+          type: TYPES.DateTimeOffset
         }
       ]
+
       const sql = `
-      DECLARE @cnt INT = 0;
+      DECLARE @cnt INT = 1;
       DECLARE @baseUpn INT = 80120000 + @schoolId
-      WHILE @cnt < 60
+      BEGIN TRAN
+
+      UPDATE ${sqlService.adminSchema}.[school]
+      SET pin = '${randomPass()}', pinExpiresAt = @pinExpiresAt
+      WHERE id = ${school.id} AND pin IS NULL
+
+      WHILE @cnt <= ${totalPupils}
       BEGIN
-        INSERT ${sqlService.adminSchema}.[pupil] (school_id, foreName, lastName, gender, dateOfBirth, upn) 
-        VALUES (@schoolId, CAST(@cnt AS NVARCHAR), 'Pupil', 'M', @dateOfBirth, CAST(@baseUpn AS NVARCHAR) + CAST(@cnt AS NVARCHAR) + '1A')
-        SET @cnt = @cnt + 1;
-      END;`
+      BEGIN TRY
+
+      INSERT ${sqlService.adminSchema}.[pupil] (school_id, foreName, lastName, gender, dateOfBirth, upn, isTestAccount, pin, pinExpiresAt) 
+      VALUES (@schoolId, CAST(@cnt AS NVARCHAR), 'Pupil', 'M', @dateOfBirth, CAST(@baseUpn AS NVARCHAR) + CAST(@cnt AS NVARCHAR) + '1A', 1,RIGHT('0000'+CAST(@cnt as NVARCHAR), 4), @pinExpiresAt)
+
+      INSERT ${sqlService.adminSchema}.[check] (pupil_id, checkwindow_id, checkform_id, checkstatus_id, islivecheck)
+      VALUES (scope_identity(), 1, 1, 1, 0)
+
+      END TRY
+      BEGIN CATCH
```

Comment:
Not much happening in the catch. Should it rollback?

Reply:
It was being used instead of `insert if not exists` to catch the duplicate error and move on, but it may not be needed any more now the pupil pins have been removed.